gnocchi/indexer/sqlalchemy.py

```diff
@@ -897,12 +897,29 @@ def create_resource(self, resource_type, id,
             return r
 
     @retry_on_deadlock
-    def update_resource(self, resource_type,
-                        resource_id, ended_at=_marker, metrics=_marker,
-                        append_metrics=False,
-                        create_revision=True,
+    def update_resource(self, resource_type, resource_id,
+                        ended_at=_marker, metrics=_marker,
+                        append_metrics=False, create_revision=True,
                         **kwargs):
         with self.facade.writer() as session:
+            data_to_update = kwargs.copy()
+
+            data_to_update['ended_at'] = ended_at
+            data_to_update['metrics'] = metrics
+            if create_revision:
+                resource = self.get_resource(
+                    resource_type, resource_id, with_metrics=True)
+                if not utils.is_resource_revision_needed(
+                        resource, data_to_update):
+                    LOG.info("We thought that a revision for resource "
+                             "[%s] was needed. However, after locking the "
+                             "table and checking it again, we found that it "
+                             "is not needed anymore. This is due to a "
+                             "concurrency issue that might happen. Therefore, "
+                             "no revision is going to be generated at this "
+                             "time.", data_to_update)
+                    create_revision = False
```

Comment:
i wonder if it makes sense to just move the check revision logic to https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/indexer/sqlalchemy.py#L922-L923. that way it would be behind the lock created by for_update.

that said, now it's always locking and i've no idea the impact.

Reply:
> This makes sense, it's probably very hard to test the race condition there, but do we have a test for the normal case of a new revision to verify this doesn't break normal behavior? (either unit or a gabbit)
> 
> Also please squash (cleanup) the commits into one (or more, whatever is preferred) but provide good commit messages.

Thanks for the review. I squashed the commits. Also, regarding tests, there are already gabbi tests that cover resource revisions, and all of them are passing just fine.

Reply:
@chungg I agree with you. Ideally, I Would have done that. However, I implemented this way to be less intrusive as possible in the code base. 

I can move the decision to generate revision to the update_resource method only, if you prefer.

Reply:
I'm fine with this, I'll let @chungg respond and approve if he's OK with it.